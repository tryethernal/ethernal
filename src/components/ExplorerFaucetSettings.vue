<template>
    <v-container fluid>
        <template v-if="explorer && explorer.faucet">
            <v-row>
                <v-col cols="6" class="fill-height">
                    <v-card outlined class="my-6">
                        <v-card-text>
                            <v-alert text type="error" v-if="errorMessage">{{ errorMessage }}</v-alert>
                            <v-alert text type="success" v-if="successMessage">{{ successMessage }}</v-alert>

                            <v-switch @click.prevent="toggleFaucet()" :loading="switchLoading || active === null" class="mt-1" v-model="active" inset :label="`${faucet.active ? 'Active' : 'Inactive'}`"></v-switch>
                            <strong>URL:</strong> <a :href="`//${mainExplorerDomain}/faucet`" target="_blank">https://{{ mainExplorerDomain }}/faucet</a><br>
                            <strong>Address:</strong> <Hash-Link :type="'address'" :hash="faucet.address" :fullHash="true" :withName="false" /><br>
                            <strong>Balance:</strong>&nbsp;
                            <template v-if="balance">{{ balance | fromWei('ether', explorer.token) }}</template>
                            <template v-else><i>Fetching...</i></template>

                            <v-divider class="mt-4 mb-6"></v-divider>

                            <v-form :disabled="!faucet.active || switchLoading || loading" @submit.prevent="updateFaucet()" v-model="valid">
                                <v-text-field
                                    dense
                                    outlined
                                    required
                                    :rules="[
                                        v => !!v || 'Amount is required',
                                        v => parseFloat(v) > 0 || 'Amount needs to be greater than 0'
                                    ]"
                                    type="number"
                                    hint="Amount to send for each request"
                                    persistent-hint
                                    v-model="amount"
                                    :suffix="`${explorer.token || 'ETH'}`"
                                    label="Drip Amount"></v-text-field>
                                <v-text-field
                                    class="mt-2"
                                    dense
                                    outlined
                                    required
                                    :rules="[
                                        v => !!v || 'Interval is required',
                                        v => parseFloat(v) > 0 || 'Interval needs to be greater than 0'
                                    ]"
                                    type="number"
                                    hint="Minimum time between requests for each address"
                                    v-model="interval"
                                    persistent-hint
                                    suffix="hours"
                                    label="Interval Between Drips"></v-text-field>
                                <v-card-actions class="pr-0 pb-0">
                                    <v-spacer></v-spacer>
                                    <v-btn :loading="loading" color="primary" :disabled="!valid || !faucet.active || switchLoading" type="submit">Update</v-btn>
                                </v-card-actions>
                            </v-form>
                        </v-card-text>
                    </v-card>
                    <h4 class="error--text">Danger Zone</h4>
                    <Explorer-Faucet-Settings-Danger-Zone @delete="loadExplorer" :explorerId="explorer.id" :faucetId="faucet.id" />
                </v-col>
            </v-row>
        </template>
        <template v-else-if="explorer">
            <v-card outlined>
                <Create-Explorer-Faucet-Modal ref="createExplorerFaucetModal" />
                <v-card-text>
                    <v-row>
                        <v-col align="center">
                            <v-icon style="opacity: 0.25;" size="200" color="primary lighten-1">mdi-faucet</v-icon>
                        </v-col>
                    </v-row>
                    <v-row>
                        <v-spacer></v-spacer>
                        <v-col cols="6" class="text-body-1">
                            Setup your faucet in two easy steps:
                            <ol>
                                <li>Enter the drip amount, and the interval between requests.</li>
                                <li>Top-up your assigned faucet address with test tokens.</li>
                            </ol>
                            <br>
                            Having a faucet integrated in your explorer will make it easier for your users to
                            request test tokens, and save you time by having one less tool to build & maintain.
                        </v-col>
                        <v-spacer></v-spacer>
                    </v-row>
                </v-card-text>
                <v-card-actions class="mb-4">
                    <v-spacer></v-spacer>
                    <v-btn :loading="loading" color="primary" @click="openCreateExplorerFaucetModal">Setup your Faucet</v-btn>
                    <v-spacer></v-spacer>
                </v-card-actions>
            </v-card>
        </template>
        <template v-else>
            <v-card outlined>
                <v-skeleton-loader type="list-item-three-line"></v-skeleton-loader>
            </v-card>
        </template>
    </v-container>
</template>

<script>
const ethers = require('ethers');
import store from '../plugins/store';
import CreateExplorerFaucetModal from './CreateExplorerFaucetModal';
import ExplorerFaucetSettingsDangerZone from './ExplorerFaucetSettingsDangerZone';
import HashLink from './HashLink';
import FromWei from '../filters/FromWei.js';

export default {
    name: 'ExplorerFaucetSettings',
    props: ['explorerId', 'sso'],
    components: {
        CreateExplorerFaucetModal,
        ExplorerFaucetSettingsDangerZone,
        HashLink
    },
    filters: {
        FromWei
    },
    data: () => ({
        loading: false,
        switchLoading: false,
        active: null,
        successMessage: null,
        errorMessage: null,
        valid: false,
        explorer:  null,
        faucet: null,
        balance: null,
        pusherUnsubscribe: null
    }),
    mounted() {
        this.loadExplorer();
        this.pusherUnsubscribe = this.pusher.onNewTransaction(data => {
            if (data.from == this.faucet.address)
                this.refreshFaucetBalance();
        }, this);
    },
    destroyed() {
        if (this.pusherUnsubscribe)
            this.pusherUnsubscribe();
    },
    methods: {
        toggleFaucet() {
            this.switchLoading = true;
            const fn = this.active ? 'activateFaucet' : 'deactivateFaucet';
            this.server[fn](this.faucet.id)
                .then(() => this.faucet.active = this.active)
                .catch(error => this.errorMessage = error.response && error.response.data || 'Error while updating faucet. Please retry.')
                .finally(() => this.switchLoading = false);
        },
        refreshFaucetBalance() {
            if (this.faucet && this.faucet.id)
                this.server.getFaucetBalance(this.faucet.id)
                    .then(({ data }) => this.balance = data.balance)
                    .catch(console.log);
        },
        updateFaucet() {
            this.loading = true;
            this.errorMessage = null;
            this.successMessage = null;
            this.server.updateFaucet(this.faucet.id, this.faucet.amount, this.faucet.interval)
                .then(() => this.successMessage = 'Settings updated.')
                .catch(error => this.errorMessage = error.response && error.response.data || 'Error while updating faucet. Please retry.')
                .finally(() => this.loading = false);
        },
        loadExplorer() {
            this.server.getExplorer(this.explorerId)
                .then(({ data }) => {
                    this.explorer = data;
                    this.faucet = data.faucet;
                    if (this.faucet) {
                        this.active = data.faucet.active;
                        this.refreshFaucetBalance();
                    }
                })
                .catch(console.log);
        },
        openCreateExplorerFaucetModal() {
            this.$refs.createExplorerFaucetModal.open({
                explorerId: this.explorerId,
                token: this.explorer.token || 'ETH'
            })
            .then(updated => {
                if (updated) {
                    this.explorer = null;
                    this.loadExplorer();
                }
            })
        }
    },
    computed: {
        mainExplorerDomain() {
            if (!this.explorer)
                return null;
            return this.explorer && this.explorer.domains && this.explorer.domains.length ?
                this.explorer.domains[0].domain :
                this.explorer.domain;
        },
        amount: {
            get() {
                return ethers.utils.formatUnits(this.faucet.amount).toString();
            },
            set(val) {
                this.faucet.amount = ethers.utils.parseUnits(val).toString();
            }
        },
        interval: {
            get() {
                return parseFloat(this.faucet.interval) / 60;
            },
            set(val) {
                this.faucet.interval = parseFloat(val) * 60;
            }
        }
    },
    watch: {
        faucet() {
            store.dispatch('updateFaucetSettings', this.faucet);
        }
    },
}
</script>
