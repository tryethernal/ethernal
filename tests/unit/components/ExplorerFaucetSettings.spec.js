import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

import ExplorerFaucetSettings from '@/components/ExplorerFaucetSettings.vue';

beforeEach(() => jest.clearAllMocks());

describe('ExplorerFaucetSettings.vue', () => {
    let helper;
    const stubs = ['Create-Explorer-Faucet-Modal', 'Explorer-Faucet-Settings-Danger-Zone', 'Hash-Link'];

    beforeEach(() => {
        helper = new MockHelper();
    });

    it('Should display faucet intro', async () => {
        jest.spyOn(helper.mocks.server, 'getExplorer').mockResolvedValueOnce({ data: { faucet: null }});

        const wrapper = helper.mountFn(ExplorerFaucetSettings, {
            stubs,
            propsData: {
                explorerId: 1
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display active faucet UI', async () => {
        jest.spyOn(helper.mocks.server, 'getExplorer').mockResolvedValueOnce({ data: {
            token: 'ETL',
            domain: 'ethernal.com',
            domains: [],
            faucet: {
                id: 1,
                address: '0x4150e51980114468aa8309bb72f027d8bff41353',
                amount: '100000000000000000',
                interval: 24 * 60,
                active: true
            }
        }});
        jest.spyOn(helper.mocks.server, 'getFaucetBalance').mockResolvedValueOnce({ data: { balance: '90000000000000000' }});

        const wrapper = helper.mountFn(ExplorerFaucetSettings, {
            stubs,
            propsData: {
                explorerId: 1
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('SHould display inactive faucet UI', async () => {
        jest.spyOn(helper.mocks.server, 'getExplorer').mockResolvedValueOnce({ data: {
            token: 'ETL',
            domain: 'ethernal.com',
            domains: [],
            faucet: {
                id: 1,
                address: '0x4150e51980114468aa8309bb72f027d8bff41353',
                amount: '100000000000000000',
                interval: 24 * 60,
                active: false
            }
        }});
        jest.spyOn(helper.mocks.server, 'getFaucetBalance').mockResolvedValueOnce({ data: { balance: '90000000000000000' }});

        const wrapper = helper.mountFn(ExplorerFaucetSettings, {
            stubs,
            propsData: {
                explorerId: 1
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });
});
