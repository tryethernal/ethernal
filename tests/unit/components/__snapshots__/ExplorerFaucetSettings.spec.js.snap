// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExplorerFaucetSettings.vue SHould display inactive faucet UI 1`] = `
<div class="container container--fluid">
  <div class="row">
    <div class="fill-height col col-6">
      <div class="my-6 v-card v-sheet v-sheet--outlined theme--light">
        <div class="v-card__text">
          <!---->
          <!---->
          <div class="v-input mt-1 theme--light v-input--selection-controls v-input--switch v-input--switch--inset">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-input--selection-controls__input"><input aria-checked="false" id="input-40" role="switch" type="checkbox" aria-disabled="false" value="">
                  <div class="v-input--selection-controls__ripple"></div>
                  <div class="v-input--switch__track theme--light"></div>
                  <div class="v-input--switch__thumb theme--light">
                    <transition-stub name="fab-transition" mode="out-in"></transition-stub>
                  </div>
                </div><label for="input-40" class="v-label theme--light" style="left: 0px; position: relative;">Inactive</label>
              </div>
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div> <strong>URL:</strong> <a href="//ethernal.com/faucet" target="_blank">https://ethernal.com/faucet</a><br> <strong>Address:</strong>
          <hash-link-stub fullhash="true" hash="0x4150e51980114468aa8309bb72f027d8bff41353" type="address"></hash-link-stub><br> <strong>Balance:</strong>&nbsp;
          0.09 ETL
          <hr role="separator" aria-orientation="horizontal" class="mt-4 mb-6 v-divider theme--light">
          <form novalidate="novalidate" class="v-form">
            <div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-disabled v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <fieldset aria-hidden="true">
                    <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                  </fieldset>
                  <div class="v-text-field__slot"><label for="input-47" class="v-label v-label--active v-label--is-disabled theme--light" style="left: 0px; position: absolute;">Drip Amount</label><input required="required" disabled="disabled" id="input-47" type="number">
                    <div class="v-text-field__suffix">ETL</div>
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                      <div class="v-messages__message">Amount to send for each request</div>
                    </transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
            <div class="v-input mt-2 v-input--is-label-active v-input--is-dirty v-input--is-disabled v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <fieldset aria-hidden="true">
                    <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                  </fieldset>
                  <div class="v-text-field__slot"><label for="input-50" class="v-label v-label--active v-label--is-disabled theme--light" style="left: 0px; position: absolute;">Interval Between Drips</label><input required="required" disabled="disabled" id="input-50" type="number">
                    <div class="v-text-field__suffix">hours</div>
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                      <div class="v-messages__message">Minimum time between requests for each address</div>
                    </transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
            <div class="v-card__actions pr-0 pb-0">
              <div class="spacer"></div> <button type="submit" disabled="disabled" class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">Update</span></button>
            </div>
          </form>
        </div>
      </div>
      <h4 class="error--text">Danger Zone</h4>
      <explorer-faucet-settings-danger-zone-stub faucetid="1"></explorer-faucet-settings-danger-zone-stub>
    </div>
  </div>
</div>
`;

exports[`ExplorerFaucetSettings.vue Should display active faucet UI 1`] = `
<div class="container container--fluid">
  <div class="row">
    <div class="fill-height col col-6">
      <div class="my-6 v-card v-sheet v-sheet--outlined theme--light">
        <div class="v-card__text">
          <!---->
          <!---->
          <div class="v-input mt-1 v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--switch v-input--switch--inset primary--text">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-input--selection-controls__input"><input aria-checked="true" id="input-18" role="switch" type="checkbox" aria-disabled="false" value="">
                  <div class="v-input--selection-controls__ripple primary--text"></div>
                  <div class="v-input--switch__track theme--light primary--text"></div>
                  <div class="v-input--switch__thumb theme--light primary--text">
                    <transition-stub name="fab-transition" mode="out-in"></transition-stub>
                  </div>
                </div><label for="input-18" class="v-label theme--light" style="left: 0px; position: relative;">Active</label>
              </div>
              <div class="v-messages theme--light primary--text">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div> <strong>URL:</strong> <a href="//ethernal.com/faucet" target="_blank">https://ethernal.com/faucet</a><br> <strong>Address:</strong>
          <hash-link-stub fullhash="true" hash="0x4150e51980114468aa8309bb72f027d8bff41353" type="address"></hash-link-stub><br> <strong>Balance:</strong>&nbsp;
          0.09 ETL
          <hr role="separator" aria-orientation="horizontal" class="mt-4 mb-6 v-divider theme--light">
          <form novalidate="novalidate" class="v-form">
            <div class="v-input v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <fieldset aria-hidden="true">
                    <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                  </fieldset>
                  <div class="v-text-field__slot"><label for="input-25" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Drip Amount</label><input required="required" id="input-25" type="number">
                    <div class="v-text-field__suffix">ETL</div>
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                      <div class="v-messages__message">Amount to send for each request</div>
                    </transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
            <div class="v-input mt-2 v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <fieldset aria-hidden="true">
                    <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                  </fieldset>
                  <div class="v-text-field__slot"><label for="input-28" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Interval Between Drips</label><input required="required" id="input-28" type="number">
                    <div class="v-text-field__suffix">hours</div>
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                      <div class="v-messages__message">Minimum time between requests for each address</div>
                    </transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
            <div class="v-card__actions pr-0 pb-0">
              <div class="spacer"></div> <button type="submit" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">Update</span></button>
            </div>
          </form>
        </div>
      </div>
      <h4 class="error--text">Danger Zone</h4>
      <explorer-faucet-settings-danger-zone-stub faucetid="1"></explorer-faucet-settings-danger-zone-stub>
    </div>
  </div>
</div>
`;

exports[`ExplorerFaucetSettings.vue Should display faucet intro 1`] = `
<div class="container container--fluid">
  <div class="v-card v-sheet v-sheet--outlined theme--light">
    <create-explorer-faucet-modal-stub></create-explorer-faucet-modal-stub>
    <div class="v-card__text">
      <div class="row">
        <div align="center" class="col"><i aria-hidden="true" class="v-icon notranslate mdi mdi-faucet theme--light primary--text text--lighten-1" style="font-size: 200px; opacity: 0.25;"></i></div>
      </div>
      <div class="row">
        <div class="spacer"></div>
        <div class="text-body-1 col col-6">
          Setup your faucet in two easy steps:
          <ol>
            <li>Enter the drip amount, and the interval between requests.</li>
            <li>Top-up your assigned faucet address with test tokens.</li>
          </ol> <br>
          Having a faucet integrated in your explorer will make it easier for your users to
          request test tokens, and save you time by having one less tool to build &amp; maintain.
        </div>
        <div class="spacer"></div>
      </div>
    </div>
    <div class="v-card__actions mb-4">
      <div class="spacer"></div> <button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">Setup your Faucet</span></button>
      <div class="spacer"></div>
    </div>
  </div>
</div>
`;
